add floating window component

Duffscs opened the pull request:
The feature isn't actually finished. it lacks z-index management for windows. 
I'd be happy to get some advice or even have someone else take care of it. 
I tried something with a stack but it wasn't very glorious. I think I'm missing some notion of how state and derive work in van. In the end, I removed that part and only kept the functional part, window creation. 
It's probably not the best way of doing things I've chosen, but I'm open to suggestions if you think we can do better.
#132

Tao-VanJS reviewed (commented):
Thank you so much for your quick work! I left a few minor comments.

Regarding `z-index`. I am thinking of a simple solution:
1. We allow users to specify the `z-index` on the callsites.
2. If the users left `z-index` unspecified, the default value can be `1, 2, 3, ...`. i.e.: incrementing by 1 every time when a new window is created. This will have the effect that newly created windows always in the front of the stack than existing windows, which might be the desirable behavior for most of the cases.

WDYT?

Duffscs commented on components/van-ui.ts:
```diff
@@ -499,3 +499,188 @@ export const Banner = (
+      x.val += e.clientX - startX.val
+      y.val += e.clientY - startY.val
+      startX.val = e.clientX
+      startY.val = e.clientY
```
No cause, If you drag the window in the middle it will move the left corner to your mouse.

Duffscs commented on components/examples/window/index.html:
I wasn't able to understand how to correctly contribute.
my workflow was
```
 .../van/components $ npm run build && yes | cp -r dist/* examples/window/node_modules/vanjs-ui/dis
 ```
 and
 ```
 .../van/components/examples/window $ npm run build && npm run preview
 ```
it's rly uneffective
If you can provide a guide to contribute somewhere it could be cool

Duffscs commented:
> Thank you so much for your quick work! I left a few minor comments.
> 
> Regarding `z-index`. I am thinking of a simple solution:
> 
>     1. We allow users to specify the `z-index` on the callsites.
> 
>     2. If the users left `z-index` unspecified, the default value can be `1, 2, 3, ...`. i.e.: incrementing by 1 every time when a new window is created. This will have the effect that newly created windows always in the front of the stack than existing windows, which might be the desirable behavior for most of the cases.
> 
> 
> WDYT?

what do you mean by callsites ?
for each created window specifying z-index or specifying a default z-index for the library ?
like FloatingWindow.zIndex = 10 and when creating a window it will do FloatingWindow.zIndex + stack.length  + 1

I think we need a bring front feature, could be a function provided by the library and fire by the user, or an automatic function trigger when you click and a window which is not on the front?

Tao-VanJS commented:
> > Thank you so much for your quick work! I left a few minor comments.
> > Regarding `z-index`. I am thinking of a simple solution:
> > ```
> > 1. We allow users to specify the `z-index` on the callsites.
> > 
> > 2. If the users left `z-index` unspecified, the default value can be `1, 2, 3, ...`. i.e.: incrementing by 1 every time when a new window is created. This will have the effect that newly created windows always in the front of the stack than existing windows, which might be the desirable behavior for most of the cases.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > WDYT?
> 
> what do you mean by callsites ? for each created window specifying z-index or specifying a default z-index for the library ? like FloatingWindow.zIndex = 10 and when creating a window it will do FloatingWindow.zIndex + stack.length + 1
> 
> I think we need a bring front feature, could be a function provided by the library and fire by the user, or an automatic function trigger when you click and a window which is not on the front?

I see your point. I was also thinking about this too: when a window is clicked, we do normally expect the window being clicked to be brought to the top.

How about the simplified mechanism:
1. The `z-index` is a monotonically increasing integer sequence: 1, 2, 3, ...
2. Whenever a new window is created, or clicked, we assign the `z-index` of the window to the next integer in the sequence. This ensures the recently created/clicked window to be always on the top.

In this way we lose the ability to programmatically re-order the window stack. But I guess it's probably okay to not adding it in the first version. If we do want to add the functionality in the future, we can allow users to pass-in a state object for the `z-index` so that it can be controlled externally.

Let me know your thought about this mechanism.

Tao-VanJS commented on components/van-ui.ts:
```diff
@@ -499,3 +499,188 @@ export const Banner = (
+          height: "1rem"
+        }),
+        onmousedown: onMouseDown,
+      }),
```
Actually, I don't quite understand the purpose of this `<div>` element. When I tried to remove the element, nothing seems to be effected. Is it redundant or there is something you intended to do?

Duffscs replied:
I we remove this div this window is not draggable when your title bar is hidden.
if remove "Window with Tabs with custom close button" can be drag. its useless most of the time, if you want we can add a boolean titlebar? and when its false we add this div to allow drag.

Tao-VanJS replied:
Can we show this `<div>` only when `!title` is true, and show the `<header>` below only when `title` is true?

Tao-VanJS commented on components/examples/window/src/main.ts:
```diff
@@ -0,0 +1,178 @@
+          tabButtonActiveColor: "white",
+          tabButtonBorderStyle: "none",
+          tabButtonRowStyleOverrides: {
+            "background-color": "lightblue",
```
I guess the `background-color` can be specified via `tabButtonRowColor` property?

Tao-VanJS commented on components/examples/window/src/main.ts:
```diff
@@ -0,0 +1,178 @@
+        }),
+        class: "custom-cross",
+        onclick: () => closed.val = true,
+      }, "×"),
```
Is this element duplicate with `closeCross: true` above?

Tao-VanJS replied:
Let remove the `closeCross: true` line since it doesn't take effect. Otherwise it might confuse the reader (they might think it's `closeCross: true` control whether the cross button to show or not).

Duffscs replied:
Oh, my bad, it was a remain of a paste. Yes it's useless

Tao-VanJS commented on components/examples/window/src/main.ts:
```diff
@@ -0,0 +1,178 @@
+  const width = van.state(500)
+  const height = van.state(300)
+
+  van.add(document.body, FloatingWindow({
```
This example is great! However, there is a minor issue: The height of the cross button is not the same as the title bar of the Tabs. As a result, the cross button appears a little bit below the center: 
![image](https://github.com/vanjs-org/van/assets/132854319/36f6ecac-f07b-4ece-a65e-010109655b9d)

I will appreciate if you can get it fixed.

Duffscs replied:
done in 1d11021

Tao-VanJS reviewed (commented):
One more thing: we probably need to add a new section in the [README.md](https://github.com/vanjs-org/van/blob/main/components/README.md) file for the `FloatingWindow` component.

Let me know if you want to add it. If not, you can leave your PR there and I can take care of the documentation.

Duffscs commented:
To improve readability
I moved non-dynamic style a unique stylesheet.
I moved dynamic style in a another dedicated stylesheet.
and state dependant style are directly use in component.style attribute

if you think it's unnecessary I can do it like other component, using variables. I did it like that because you can have a lot of window it'll result on a lot of duplicated css.

Duffscs commented:
I'll take care of z-index soon

Tao-VanJS commented:
> To improve readability I moved non-dynamic style in a unique stylesheet. I moved dynamic style in a another dedicated stylesheet. and state dependant style are directly use in component.style attribute
> 
> if you think it's unnecessary I can do it like other component, using variables. I did it like that because you can have a lot of window it'll result on a lot of duplicated css.

Since you're the contributor of this component you can choose the style that you like :-). Personally, while implementing VanUI, I'm trying to make each component hermetic, thus reducing the side effect of the component as much as possible. This is why I was choosing inline styles over CSS injection as much as I can (for `Tabs` component, I have to inject CSS as `:hover` selector is not supported in inline styles).

Tao-VanJS commented:
> I'll take care of z-index soon

Hi @Duffscs,

A gentle nudge, is the implementation for `z-index` something you're planning to work on?

Duffscs commented:
Hi @Tao-VanJS, 

In the last few days I've been finishing an important AI assignment, but I'm going to have some free time again and get back to it.

(assignment is due monday)

Tao-VanJS commented:
@all-contributors please add @Duffscs for code and plugin
